### ISSUE TITLE
Improve message “The banner role is unnecessary for element header”

### ISSUE THREAD

**Issue by cescarment** [validator/validator#270]
Hello,

When the HTML Header tag is not a descendant of an article or section element, the element has a strong native semantics and the default implicit ARIA semantics is the banner role.
In this case (not a descendant of an article or section element), the validator works fine. Explicitly specify a banner role causes a warning, and explicitly specify an other role (except group and presentation) causes an error.

In the case of the HTML Header tag is a descendant of an article or section element, the element don't have a strong native semantics and don't have a default implicit ARIA semantics. So we can expect that the author can specify explicitly a role attribute. But the validator causes a warning for banner role, and an error for others roles (except group and presentation). Exactly the same behavior in the case of the HTML Header tag is not a descendant of an article or section element ...

**Comment by sideshowbarker** [comment]
As far as I understand it, the checker is conforming to the relevant requirements for this in the ARIA in HTML spec https://www.w3.org/TR/html-aria/

cc @stevefaulkner

**Comment by sideshowbarker** [comment]
@stevefaulkner can you please take a look at this and let me know if any change is needed.

**Comment by stevefaulkner** [comment]
@cescarment @sideshowbarker While it is correct that headers/footers not scoped to body don't have strong native semantics, the warning still appears because we want to discourage developers from adding banner/contentinfo roles to these elements. Reason being that addition as a matter of course will create unecessary semantic verbosity for users. So the general rule is don't do it, but the validator throws a warning not an error, as there may be cases where the addition of such roles to these elements may be appropriate. This article may be helpful https://www.paciellogroup.com/blog/2015/08/short-note-on-html-conformance-checking/

**Comment by cedricescarment** [comment]
For the warning on banner role on a header not scoped to body : thanks for the explanations why it's not recommanded, but the explanations of the warning are not clear. The warning is : 

> Warning: the banner role is unnecessary for element header.

So i understand the validator consider that the banner element has the strong native semantics banner ?

```html
<!doctype html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<title>My test page</title>
	</head>
	<body>
		<h1>My test page</h1>
		<section>
			<header role="banner">
				<p>my content info</p>
			</header>
			<h2>One section</h2>
			<p>my section</p>
		</section>
	</body>
</html>

**Comment by sideshowbarker** [comment]
> So i understand the validator consider that the banner element has the strong native semantics banner ?

Assuming you mean the `header` element (instead of `banner` element), then yes. See https://www.w3.org/TR/html-aria/#header

**Comment by cedricescarment** [comment]
Yes i mean the header element, but the documentation explains header element has no corresponding role if it's a descendant of section element. I still think the sentence of the warning of my exemple is not correct. It dont explain the argument of Steve Faulkner for the reason of the warning, it's the warning in the case of a header element not descendant of section or article.

**Comment by stevefaulkner** [comment]
> I still think the sentence of the warning of my example is not correct.

I agree that the current warning is unclear, @sideshowbarker what's the best way for me to provide advice on a more meaningful message for this?

**Comment by sideshowbarker** [comment]
> what's the best way for me to provide advice on a more meaningful message for this?

If you mean the warning “The `banner` role is unnecessary for element `header`” and you want to provide a patch, the source is here:

https://github.com/validator/validator/blob/master/src/nu/validator/checker/schematronequiv/Assertions.java#L2391

But not that any change for this to the source in its current form is also going to affect the text of the warnings for the case of a `role` attribute used with any other elements that have implicit semantics—`article`, `aside`, `button`, `h1`-`h6`, `img`, `ol`, `ul`, etc.

But I could change the code to special-case `header` with a more-specific warning. Just let me know what that specific warning should say.

**Comment by j9t** [comment]
@stevefaulkner, @sideshowbarker, a couple of years late but also happy to help if useful 🙂

**Comment by sideshowbarker** [comment]
Made an update for this in https://github.com/validator/validator/commit/272b12e6f68088142a18bd7126f1b2aa4ef4d3a0 and pushed it to https://validator.w3.org/nu/
